Sources/TSCBasic/Path.swift

```diff
@@ -865,53 +366,37 @@ extension AbsolutePath {
     ///
     /// This method is strictly syntactic and does not access the file system
     /// in any way.  Therefore, it does not take symbolic links into account.
-    public func relative(to base: AbsolutePath) -> RelativePath {
-        let result: RelativePath
-        // Split the two paths into their components.
-        // FIXME: The is needs to be optimized to avoid unncessary copying.
-        let pathComps = self.components
-        let baseComps = base.components
-
-        // It's common for the base to be an ancestor, so try that first.
-        if pathComps.starts(with: baseComps) {
-            // Special case, which is a plain path without `..` components.  It
-            // might be an empty path (when self and the base are equal).
-            let relComps = pathComps.dropFirst(baseComps.count)
-#if os(Windows)
-            result = RelativePath(relComps.joined(separator: "\\"))
-#else
-            result = RelativePath(relComps.joined(separator: "/"))
-#endif
-        } else {
-            // General case, in which we might well need `..` components to go
-            // "up" before we can go "down" the directory tree.
-            var newPathComps = ArraySlice(pathComps)
-            var newBaseComps = ArraySlice(baseComps)
-            while newPathComps.prefix(1) == newBaseComps.prefix(1) {
-                // First component matches, so drop it.
-                newPathComps = newPathComps.dropFirst()
-                newBaseComps = newBaseComps.dropFirst()
-            }
-            // Now construct a path consisting of as many `..`s as are in the
-            // `newBaseComps` followed by what remains in `newPathComps`.
-            var relComps = Array(repeating: "..", count: newBaseComps.count)
-            relComps.append(contentsOf: newPathComps)
+    public func relative(to base: AbsolutePath) throws -> RelativePath {
+        var relFilePath = FilePath()
+        var filepath = filepath
 #if os(Windows)
-            result = RelativePath(relComps.joined(separator: "\\"))
-#else
-            result = RelativePath(relComps.joined(separator: "/"))
-#endif
+        /// TODO: DOS relative path may change the root.
+        if root != base.root {
+            throw PathValidationError.differentRoot(pathString, base.pathString)
         }
-        assert(base.appending(result) == self)
-        return result
+#endif
+        filepath.root = base.filepath.root
+
+        let commonAncestor = AbsolutePath(filepath).lowestCommonAncestor(with: base)!
```

Comment:
guard + precondition instead of bang?

Reply:
I don’t think callers should guard such condition🤔 Breaking is quite reasonable and necessary here.

We’d better discuss in #219 since this API change is introduced there.